DOCS-7377-vm-new-examples-kt (#1676)
mulesoft/docs-connectors: vm/2.0/modules/ROOT/pages/vm-aggregate-messages.adoc

. In Studio, drag a *VM > Listener* operation below the first flow.
. For *Connector configuration*, select the previously created `VM_Config` configuration.
. Set *Queue name* to `demo`.
. Drag a *Time based aggregator* scope to the right of the *VM Listener*.
. Set *Name* to `demoaggregator`.
. Set *Period* to `5` to aggregate the received messages until 5 seconds is reached.

=== Create the Third Flow

In this flow, an *Aggregator listener* operation listens to the messages aggregated by the *Time based aggregator* scope. The *For Each* component iterates over the messages and a *Logger* component logs the content of each message.

. In Studio, drag an *Aggregator listener* source above the second flow.
. Set *Aggregator name* to `demoaggregator`. +
. Drag a *For Each* component to the right of the *Aggregator listener* source.
. Set *Collection* to `[#payload]`.
. Drag a *Logger* component into the *For Each* component.
. Set *Message* to `[#payload]`.
. Save and run your Mule app.
. In your terminal, run the curl command `curl http://localhost:8081/`.

> === XML for the First App Example

Reviewer: Should match the example name at least partially - XML for the First Example, ditto for the second one
I think having App confuses things.

= Aggregate Messages Pending in Queue with VM Connector Example

The following examples show two scenarios on how to aggregate messages pending on a queue using Anypoint Connector for Virtual Machine (VM Connector).

> Note that VM is a queue and you can not simply perform a for each over every message pending on a queue.

Reviewer: for each operation??? - Needs something, but I don't think that "operation" is right.

Reviewer: ```suggestion
Note that VM is a queue and you cannot perform a for each over every message pending on a queue.
```

> == Create First Mule App Example

Reviewer: Can you make these heading more informative?

In this example, you use the Aggregators module to aggregate the received messages by the VM *Listener* operation and then iterate over the result of the aggregation.

image::vm-aggregate-flow-1.png[Aggregate Messages Pending in Queue first flow in Anypoint Studio canvas]

=== Create the First Flow

> In the first flow you add an *HTTP Listener* source that listens for incoming HTTP messages, and the VM *Publish* operation that publish specific content in a VM queue:

Reviewer: ```suggestion
In the first flow you add an *HTTP Listener* source that listens for incoming HTTP messages. Then you add the VM *Publish* operation that publishes specific content in a VM queue:
```

Reviewer: An alternative wording for this paragraph is:

In the first flow, an  *HTTP Listener* source listens for incoming HTTP messages. Then a VM *Publish* operation publishes specific content in a VM queue:

. In Studio, create a new Mule project.
. In *Mule Palette*, select the *HTTP > Listener* source and drag it on to the canvas. +
. On the HTTP *Listener* configuration screen, set the *Path* field to `/`.
. Click the plus sign (+) next to the *Connector configuration* field to configure a global element that can be used by all instances of the HTTP *Listener* in the app.
. Complete the required fields for the HTTP global configuration.
. Click *OK*.
. Drag a *For Each* component to the right of the HTTP *Listener* source. +
. Set *Collection* to `1 to 200`.
. Drag a *VM > Publish* operation into the *For Each* component.
. Set *Content* to `"hello world"`.
. Click the plus sign (+) next to the *Connector configuration* field to configure a global element that can be used by all instances of the VM *Publish* operation in the app.
. In the *VM Config* window, click the plus sign (+) to add a new queue.
. Set *Queue name* to `demo`.
. Click *Finish*.
. Click *OK*.

=== Create the Second Flow

> In the second flow, you add *VM Listener* operation that listens to the messages published in the VM queue. Then, the *Time based aggregator* scope aggregates the received messages until the time condition is reached.

Reviewer: Same comment as for the first flow. The "Then" doesn't work in these intros because, for example, here it can refer either to the action of adding the listener or to the actual action the listener performs. 

Is the scope called *Time based aggregator* or did you mean to say something like "the time-based aggregator"?

. In Studio, drag a *VM > Listener* operation below the first flow.
. For *Connector configuration*, select the previously created `VM_Config` configuration.
. Set *Queue name* to `demo`.
. Drag a *Time based aggregator* scope to the right of the *VM Listener*.
. Set *Name* to `demoaggregator`.
. Set *Period* to `5`. +

> To aggregate the receved messages until 5 seconds is reached.

Reviewer: "Set *Period* to `5` to aggregate..."

http://www.mulesoft.org/schema/mule/aggregators http://www.mulesoft.org/schema/mule/aggregators/current/mule-aggregators.xsd">
	<http:listener-config name="HTTP_Listener_config" basePath="/" >
		<http:listener-connection host="0.0.0.0" port="8081" />
	</http:listener-config>
	<vm:config name="VM_Config">
		<vm:connection />
		<vm:queues >
			<vm:queue queueName="demo" />
		</vm:queues>
	</vm:config>
	<flow name="demo-vm2Flow" >
		<http:listener config-ref="HTTP_Listener_config" path="/"/>
		<foreach collection="1 to 200">
			<vm:publish config-ref="VM_Config" queueName="demo" >
				<vm:content ><![CDATA[#["hello world"]]]></vm:content>
			</vm:publish>
		</foreach>
	</flow>
	<flow name="demo-vm2Flow2" >
		<aggregators:aggregator-listener aggregatorName="demoaggregator" includeTimedOutGroups="true"/>
		<foreach collection="#[payload]">
			<logger level="INFO" message="#[payload]"/>
		</foreach>
	</flow>
	<flow name="demo-vm2Flow1" >
		<vm:listener queueName="demo" config-ref="VM_Config"/>
		<aggregators:time-based-aggregator  name="demoaggregator" period="5">
		</aggregators:time-based-aggregator>
	</flow>
</mule>
----

== Create Second Mule App Example

> In this example, you use a *For Each* component to iterate over consumed messages via the VM *Consume* operation. Then the Mule app appends every received message into a variable, and for each message logs the payload content. Note that in this example scenario you get slices of messages until no messages are available on the queue. When this occurs, the VM *Consume* operation throws a timeout message handled by the *On Error Continue* component and then the results are available for iteration.

Reviewer: See prior comments regarding heading and "Then" wording. 

"and the for each message"  Still not sure what to do with "for each."

. Save and run your Mule app.
. In your terminal, run the curl command `curl http://localhost:8081/`.

> === XML for Second App Example

Reviewer: for the Second...
